keps/sig-api-machinery/2161-apiserver-default-labels/README.md

```diff
@@ -0,0 +1,408 @@
+- closed failures which break on old clients (making networkpolicy api evolution very hard)
+- open failures with lots of warnings and docs about how old clients might overinterpret the allowed connectivity for a policy
+
+Both of these are impractical due to the nature of security tools, which need to be robust and explicit.
+
+
+Ideally, a fix here would “just work” for “back-rev” implementations of NetworkPolicy (which are out-of-tree - generally implemented by cni plugins) and would not introduce new surprises for users… But in practicality this is not the case because, by adding a new selection data structure, the logical inclusion/exclusion properties of namespace and pod selectors become drastically different, as shown in the table below:
+
+
+
+### Goals
+
+- Add an immutable default label to ALL namespaces, as the namespace name so this can be used as selector by arbitrary components.
+
+### Non-Goals
+
+It’s also worth noting that there are at least 2 other NamespaceSelector APIs in k/k - Validating and Mutating webhooks.  They will inevitably need similar API evolution, but we don't  need to do that as part of this KEP.
+
+## Proposal
+
+We propose to label of all namespaces, by default, with a reserved label (i.e. “kubernetes.io/metadata.namespace” or some such), to allow easy selection.
+If this label is missing, it is added by the apiserver
+If the label is incorrect or changed, the apiserver fails to validate the object
+
+
+### User Stories (Optional)
+
+See https://github.com/kubernetes/enhancements/pull/2113/files for relevant user stories.
+
+### Notes/Constraints/Caveats (Optional)
+
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+
+This can be implemented by modifying the defaults.go and validation.go components for the namespace API, i.e., in the defaults.go file for api/core:
+
+```
+func SetDefaults_Namespace(obj *v1.Namespace) {
+if obj.Labels == nil {
+   		obj.Labels = map[string]string{}
+}
+if _, found := obj.Labels[“kubernetes.io/metadata.name”]; !found {
+   		obj.Labels[“kubernetes.io/metadata.name”] = obj.Name
+}
+}
+```
+And in the validation.go, we implement the guarantee of this namespace’s value being constant WRT namespace name (which is immutable).
+
+```
+func ValidateNamespace(namespace *core.Namespace) field.ErrorList {
+	// ...
+   	// Check if namespace.Labels[“kubernetes.io/metadata.name”] == namespace.Name
+// if not add an error
+   	// ...
+return allErrs
+```
+
+This solves not just NetworkPolicy, but all such namespace selector APIs without any disruption or version skew problems.
+
+### Test Plan
+
```

Comment:
We need a plan here for testing - we can do a unit test that loops over namespaces, and does a select by this label and verifies the results, plus the obvious tests for trying to modify it and failing

Reply:
defaults.go populates on read, but I don't know about watch.  We need to test that WATCH on objects that already exist and DO NOT have the label actually written will work, and if not we need to find a plan

Reply:
why would a watch *not* work?  Its just the mutation of a label, right ? 
I suppose someone could write a watch that selected `! kubernetes.io/metadata.name` but that would 
be a very odd circumstance , i.e. unless someone went into the future and wanted to avoid auto labelled namespaces after reading this KEP ? :) 

or i might be missing a more obvious breaking scenario?

Reply:
I thought maybe it would not work, but I just tried it and watch does seem to work correctly.